Service to Service Auth using JWT (#21)
CoreWCF/samples: Scenarios/Authentication/Service-to-service-JWT-using-Azure-AD/Readme.md

> # Service to Service authentication using a JWT

Reviewer: ```suggestion
# Service-to-Service Authentication Using a JWT
```

## Introduction

This sample shows how a client can authenticate with a CoreWCF service using JWT-based authentication. In this specific sample, it is using Azure Active Directory as the identity provider, and the calling app is another service. The same pattern will apply to other JWT based authentication providers or to end-user rather than service to service authentication.

## What is a JWT

> [Jason Web Token](https://en.wikipedia.org/wiki/JSON_Web_Token) is a way to support federated authentication and authorization of http requests. The client makes a call to an authentication server passing credentials and identifying information about the resource they wish to access. If authentication succeeds and the resource is authorized, then they are handed back a signed token, which includes information about the resource and rights. They then pass that token to the service, which can verify its integrity, and then trust the claims it specifies.

Reviewer 2: Please add reference to rfc 7519 https://www.rfc-editor.org/rfc/rfc7519

- At design time, the [User Secrets](https://docs.microsoft.com/aspnet/core/security/app-secrets) feature will create a configuration overlay file that can be used to store secrets. In Visual Studio, right click on the project and choose *Manage User Secrets* to create and open the overlay file. Client Secrets should be stored in this file during development.
- At runtime, Environment Variables will overlay over config properties. Connection strings and secrets can be placed there and then accessed via configuration. Hosting environments have support for safely storing secrets and supplying them at runtime. Eg [App Service](https://docs.microsoft.com/azure/app-service/configure-common?tabs=portal#configure-app-settings) or [Azure Container Apps](https://docs.microsoft.com/azure/container-apps/manage-secrets?tabs=azure-cli)

The appsettings.json files in the projects include the names of the config parameters that are required for the AAD authentication. They can be stored in User Secrets for additional security during development.

# Using JWT with WCF Services

> The WS-* specifications which define the SOAP protocol and form the basis for WCF were developed long before JWT came onto the scene as the preferred form of web authentication. For this reason the WCF client APIs don't include direct support for JWT-based authentication or Authorization. However, JWT is implemented over http by supplying the token as a base64 encoded string as the authenticate header. So these samples add that header and validate it as part of the service call.

Author: yes

> The WS-* specifications which define the SOAP protocol and form the basis for WCF were developed long before JWT came onto the scene as the preferred form of web authentication. For this reason the WCF client APIs don't include direct support for JWT-based authentication or authorization. However, JWT is implemented over http by supplying the token as a base64-encoded string as the `Authorization` header. These samples add that header and validate it as part of the service call.

Reviewer 2: http header is: `Authorization: Bearer <access_token>`

## Azure AD configuration
Within the Azure AD Tenent used for the authentication, the following need to be configured. See this [Readme](https://github.com/Azure-Samples/active-directory-dotnetcore-daemon-v2/tree/master/4-Call-OwnApi-Pop) for instructions.
- App registration for the Tile Service
  - With an App role of `AccessTileBag`
- App registration of the Word Game app
  - With a client secret to identify the app
  - Granted the API Permission of `AccessTileBag` for the Tile Service API

## Tile Service (Server app)
TileService is a CoreWCF Server app. It supports one API defined in *ITileService.cs* for `DrawTiles`. 

The app uses the Azure AD integration with ASP.NET Core to enable and perform JWT Authentication. This is done through:

> - Nuget references for `Microsoft.Identity.Web` and `Microsoft.VisualStudio.Azure.Containers.Tools.Targets`

Reviewer: ```suggestion
- Adding NuGet references for `Microsoft.Identity.Web` and `Microsoft.VisualStudio.Azure.Containers.Tools.Targets`
```

- Including the ASP.NET Core and Azure AD SDK support for JWT with

``` c#
// Asp.NET Core Authentication
builder.Services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
  .AddMicrosoftIdentityWebApi(builder.Configuration);
```

- Adding Authentication and Authorization to the ASP.NET Core pipeline with

``` c#
app.UseAuthentication();
app.UseAuthorization();
```

> - Attributing the service call with the claims that need to be present

Reviewer: Suggestion: include links to docs for things like AuthorizeRole

``` c#
[AuthorizeRole("AccessTileBag")]
public IList<GameTile> DrawTiles(int count)
{
  ...
}
```

> - The parameters for the Azure SDK to validate the JWT are provided through configuration in appsettings.json

Reviewer: ```suggestion
- Configuring the parameters for the Azure SDK to validate the JWT in appsettings.json
```

``` json
  "AzureAd": {
    "Instance": "https://login.microsoftonline.com/",
    "ClientId": "[Enter the Client Id of the service (Application ID obtained from the Azure portal), e.g. ba74781c2-53c2-442a-97c2-3d60re42f403]",
    "Domain": "[Enter the domain of your tenant, e.g. contoso.onmicrosoft.com]",
    "TenantId": "[Enter 'common', or 'organizations' or the Tenant Id (Obtained from the Azure portal. Select 'Endpoints' from the 'App registrations' blade and use the GUID in any of the URLs), e.g. da41245a5-11b3-996c-00a8-4d99re19f292]"
  }
```

## WordGame (Service Client)

The service client is included in a WebAPI app that exposes a simple HTTP GET endpoint to fetch the tiles.

To make the WCF service calls, the app uses a Service Reference client wrapper, generated from the WSDL from the Tile Service. 

The key parts of the application are:

- At startup, it calls `getAzAdJwtBlob`, which

>   - Reads AAD properties from configuration (appsetting.json) and User Secrets

Reviewer: ```suggestion
  - Reads AAD properties from configuration (appsettings.json) and User Secrets
```

  - Calls AAD to get a JWT for this specific service
- Exposes a WebAPI at `/getTiles` that will make the WCF service call and return the response as JSON
  - Exposes a WebAPI at `/` which redirects to /getTiles with a count parameter
- The `getTiles` function which

>   - Uses an `OperationContextScope` to add an http `Authorize` header with the JWT as the value

Reviewer: ```suggestion
  - Uses an `OperationContextScope` to add an http `Authorization` header with the JWT as the value
```

  - Calls the Tile Service API using the generated wrapper class

> The only AAD specific code in this sample is included in the getAzAdJWTBlob function. The same pattern can be followed to retrieve a JWT from other authentication providers, or performing user authentication etc.

Reviewer: ```suggestion
The only AAD-specific code in this sample is included in the getAzAdJWTBlob function. The same pattern can be followed to retrieve a JWT from other authentication providers, or for performing user authentication, etc.
```
